### PR TITLE
device agnostic test_accelerator/test_multigpu

### PR DESCRIPTION
**Pull request by wangshuai09**
# What does this PR do?

Part of #2122 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/accelerate/blob/main/CONTRIBUTING.md#submitting-a-pull-request-pr),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/accelerate/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/accelerate/tree/main/docs#writing-documentation---specification).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@muellerzr

### CODE REVIEW

**Comment by wangshuai09** [comment]
Verified on NPU:

```shell
root@ascend-01:/root/accelerate# RUN_SLOW=1 python -m pytest -v tests/test_accelerator.py 
=========================================================================== test session starts ===========================================================================
platform linux -- Python 3.9.18, pytest-7.4.4, pluggy-1.3.0 -- /root/miniconda/envs/torch_npu/bin/python
cachedir: .pytest_cache
rootdir: /root/accelerate
collected 22 items                                                                                                                                                        

tests/test_accelerator.py::AcceleratorTester::test_accelerator_bnb SKIPPED (test requires the bitsandbytes library)                                                 [  4%]
tests/test_accelerator.py::AcceleratorTester::test_accelerator_bnb_cpu_error SKIPPED (test requires the bitsandbytes library)                                       [  9%]
tests/test_accelerator.py::AcceleratorTester::test_accelerator_bnb_multi_device SKIPPED (test requires the bitsandbytes library)                                    [ 13%]
tests/test_accelerator.py::AcceleratorTester::test_accelerator_bnb_multi_device_no_distributed SKIPPED (test requires the bitsandbytes library)                     [ 18%]
tests/test_accelerator.py::AcceleratorTester::test_accelerator_can_be_reinstantiated PASSED                                                                         [ 22%]
tests/test_accelerator.py::AcceleratorTester::test_accelerator_cpu_flag_prepare PASSED                                                                              [ 27%]
tests/test_accelerator.py::AcceleratorTester::test_accelerator_none PASSED                                                                                          [ 31%]
tests/test_accelerator.py::AcceleratorTester::test_can_unwrap_model PASSED                                                                                          [ 36%]
tests/test_accelerator.py::AcceleratorTester::test_can_unwrap_model_fp16 PASSED                                                                                     [ 40%]
tests/test_accelerator.py::AcceleratorTester::test_env_var_device PASSED                                                                                            [ 45%]
tests/test_accelerator.py::AcceleratorTester::test_free_memory_dereferences_prepared_components PASSED                                                              [ 50%]
tests/test_accelerator.py::AcceleratorTester::test_is_accelerator_prepared PASSED                                                                                   [ 54%]
tests/test_accelerator.py::AcceleratorTester::test_mutable_states PASSED                                                                                            [ 59%]
tests/test_accelerator.py::AcceleratorTester::test_prepared_objects_are_referenced PASSED                                                                           [ 63%]
tests/test_accelerator.py::AcceleratorTester::test_save_load_model_use_pytorch PASSED                                                                               [ 68%]
tests/test_accelerator.py::AcceleratorTester::test_save_load_model_use_safetensors PASSED                                                                           [ 72%]
tests/test_accelerator.py::AcceleratorTester::test_save_load_model_with_hooks_use_pytorch PASSED                                                                    [ 77%]
tests/test_accelerator.py::AcceleratorTester::test_save_load_model_with_hooks_use_safetensors PASSED                                                                [ 81%]
tests/test_accelerator.py::AcceleratorTester::test_save_model_offload_use_pytorch PASSED                                                                            [ 86%]
tests/test_accelerator.py::AcceleratorTester::test_save_model_offload_use_safetensors PASSED                                                                        [ 90%]
tests/test_accelerator.py::AcceleratorTester::test_save_model_use_pytorch PASSED                                                                                    [ 95%]
tests/test_accelerator.py::AcceleratorTester::test_save_model_use_safetensors PASSED                                                                                [100%]

============================================================================ warnings summary =============================================================================
../miniconda/envs/torch_npu/lib/python3.9/site-packages/torch_npu/dynamo/__init__.py:18
  /root/miniconda/envs/torch_npu/lib/python3.9/site-packages/torch_npu/dynamo/__init__.py:18: UserWarning: Register eager implementation for the 'npu' backend of dynamo, as torch_npu was not compiled with torchair.
    warnings.warn(

tests/test_accelerator.py::AcceleratorTester::test_save_model_offload_use_pytorch
tests/test_accelerator.py::AcceleratorTester::test_save_model_offload_use_safetensors
  /root/miniconda/envs/torch_npu/lib/python3.9/site-packages/torch/nn/modules/module.py:2025: UserWarning: for batchnorm.weight: copying from a non-meta parameter in the checkpoint to a meta parameter in the current model, which is a no-op. (Did you mean to pass `assign=True` to assign items in the state dictionary to their corresponding key in the module instead of copying them in place?)
    warnings.warn(f'for {key}: copying from a non-meta parameter in the checkpoint to a meta '

tests/test_accelerator.py::AcceleratorTester::test_save_model_offload_use_pytorch
tests/test_accelerator.py::AcceleratorTester::test_save_model_offload_use_safetensors
  /root/miniconda/envs/torch_npu/lib/python3.9/site-packages/torch/nn/modules/module.py:2025: UserWarning: for batchnorm.bias: copying from a non-meta parameter in the checkpoint to a meta parameter in the current model, which is a no-op. (Did you mean to pass `assign=True` to assign items in the state dictionary to their corresponding key in the module instead of copying them in place?)
    warnings.warn(f'for {key}: copying from a non-meta parameter in the checkpoint to a meta '

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=============================================================== 18 passed, 4 skipped, 5 warnings in 17.18s ================================================================
```
There are four cases skipped because NPU has not yet adapted to bitsandbytes.

```shell
root@ascend-01:/root/accelerate# RUN_SLOW=1 python -m pytest -v tests/test_multigpu.py 
=========================================================================== test session starts ===========================================================================
platform linux -- Python 3.9.18, pytest-7.4.4, pluggy-1.3.0 -- /root/miniconda/envs/torch_npu/bin/python
cachedir: .pytest_cache
rootdir: /root/accelerate
collected 4 items                                                                                                                                                         

tests/test_multigpu.py::MultiDeviceTester::test_distributed_data_loop PASSED                                                                                        [ 25%]
tests/test_multigpu.py::MultiDeviceTester::test_multi_device PASSED                                                                                                 [ 50%]
tests/test_multigpu.py::MultiDeviceTester::test_multi_device_ops PASSED                                                                                             [ 75%]
tests/test_multigpu.py::MultiDeviceTester::test_pad_across_processes PASSED                                                                                         [100%]

============================================================================ warnings summary =============================================================================
../miniconda/envs/torch_npu/lib/python3.9/site-packages/torch_npu/dynamo/__init__.py:18
  /root/miniconda/envs/torch_npu/lib/python3.9/site-packages/torch_npu/dynamo/__init__.py:18: UserWarning: Register eager implementation for the 'npu' backend of dynamo, as torch_npu was not compiled with torchair.
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
================================================================ 4 passed, 1 warning in 115.71s (0:01:55) =================================================================
```

**Review by statelesshz — APPROVED** [review]

File: tests/test_accelerator.py [hunk 1]
```diff
@@ -326,12 +326,17 @@ def test_accelerator_bnb_cpu_error(self):
...
     @require_bnb
-    @require_multi_gpu
-    def test_accelerator_bnb_multi_gpu(self):
+    @require_multi_device
+    def test_accelerator_bnb_multi_device(self):
         """Tests that the accelerator can be used with the BNB library."""
         from transformers import AutoModelForCausalLM
 
-        PartialState._shared_state = {"distributed_type": DistributedType.MULTI_GPU}
+        if torch_device == "cuda":
```
Comment (muellerzr) [inline tests/test_accelerator.py:334]: This is fine for now, if we notice needing this more then we should define this somewhere before as a shared var similar to `torch_device` (but inside the test file itself)

**Review by muellerzr — APPROVED** [review]
Thanks! 🤗

File: tests/test_accelerator.py (hunk 1 above)
Comment (wangshuai09) [reply tests/test_accelerator.py:334]: That makes sense. I will try if need more later.

**Pull request merged**